### PR TITLE
[JENKINS-67188] Deadlock loading `TaskListener` vs. `StreamTaskListener`

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-67188](https://issues.jenkins-ci.org/browse/JENKINS-67188). @Vlatombe diagnosed a test failure of the form

```
org.junit.runners.model.TestTimedOutException: test timed out after 180 seconds
	at app//hudson.model.TaskListener.<clinit>(TaskListener.java:158)
	at app//jenkins.model.Jenkins.<init>(Jenkins.java:1028)
	at app//hudson.model.Hudson.<init>(Hudson.java:85)
	at app//org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:688)
	at app//org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:404)
	at …
```

where the stack trace points to https://github.com/jenkinsci/jenkins/blob/e6dabb2440d2e51156068c6ea129a0e31b49111e/core/src/main/java/hudson/model/TaskListener.java#L158 and https://github.com/jenkinsci/jenkins/blob/e6dabb2440d2e51156068c6ea129a0e31b49111e/core/src/main/java/jenkins/model/Jenkins.java#L1028 which likely implicates https://github.com/jenkinsci/jenkins/blob/e6dabb2440d2e51156068c6ea129a0e31b49111e/core/src/main/java/hudson/util/LogTaskListener.java#L50 and https://github.com/jenkinsci/jenkins/blob/e6dabb2440d2e51156068c6ea129a0e31b49111e/core/src/main/java/hudson/util/StreamTaskListener.java#L61

### Proposed changelog entries

* Jenkins startup could hang due to a deadlock in class loading.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/TaskListener.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/model/TaskListener.java:158]: nit: maybe add `@FunctionalInterface` to the interface declaration for clarity (assuming that it would not impose any additional backwards compatibility constraints).

**Review by dwnusbaum — APPROVED** [review]

**Review by Vlatombe — APPROVED** [review]

**Comment by jglick** [comment]
Test failure is legitimate. Cannot use a lambda or anonymous inner class here.

**Review by Vlatombe — APPROVED** [review]

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
